### ISSUE TITLE
impossible to type '{' on french keyboard with OS X

### ISSUE THREAD

**Issue by jbpotonnier** [Microsoft/vscode#4712]
- VSCode Version: 0.10.11
- OS Version: OS X 10.11.3

Steps to Reproduce:
1.  type alt-(
2. Nothing happen

My keyboard is a french azerty (Apple style)

It works in terminal and other editors like Emacs, but I have the same problem with Atom.
I guess it might be related to https://github.com/atom/atom-keymap/issues/35
Without this key, the editor is useless to me.

**Comment by alexandrudima** [comment]
@jbpotonnier I have just tried on a fresh install of VSCode 0.10.11 on the Mac with the French azerty keyboard layout. Since I have a physical keyboard in English I type `Alt+5` and I get a `{`.

Is it possible that you have installed some extensions that bind `alt+5`. Does this reproduce if you run VSCode without extensions (the `--disable-extensions` flag)?

**Comment by jbpotonnier** [comment]
@alexandrudima I don't think I have a special extension. I'll check tonight.

**Comment by alexandrudima** [comment]
@jbpotonnier Do you still have this issue?

**Comment by jbpotonnier** [comment]
@alexandrudima Yes it is still here.
I'm now using vscode mostly on Linux and sometime with osx, both with dvorak mapping, so I'm not really affected, but the bug is still there.

**Comment by alexandrudima** [comment]
@jbpotonnier 

To clear this, I believe you mean `AltGr+5` is not working for you (`AltGr` as in the `Alt` key to the right of the spacebar). I am also assuming you are on osx with the french azerty keyboard layout.

May I ask you to try the following:

### Check what Chrome sends our way
* Open in Chrome https://www.w3.org/2002/09/tests/keys.html
* Press `AltGr+5`
* Provide a screenshot
* Here I am doing `AltGr+5` on a Windows machine. Windows a slightly different kb layout for French, and `AltGr+5` produces `[`, but the principle is the same:
![image](https://cloud.githubusercontent.com/assets/5047891/19797705/fb0895cc-9cec-11e6-98c3-f8853e976f44.png)

```
keydown (0): [keyCode: 17][shift: false][ctrl: true][alt: false][meta: false]
keydown (1): [keyCode: 17][shift: false][ctrl: true][alt: false][meta: false]
keydown (2): [keyCode: 18][shift: false][ctrl: true][alt: true][meta: false]
keydown (3): [keyCode: 18][shift: false][ctrl: true][alt: true][meta: false]
keydown (4): [keyCode: 53][shift: false][ctrl: true][alt: true][meta: false]
keydown (5): [keyCode: 53][shift: false][ctrl: true][alt: true][meta: false]
keypress (6): [keyCode: 91][charCode: 91][character: [][shift: false][ctrl: true][alt: true][meta: false]
keypress (7): [keyCode: 91][charCode: 91][character: [][shift: false][ctrl: true][alt: true][meta: false]
keyup (8): [keyCode: 53][shift: false][ctrl: true][alt: true][meta: false]
keyup (9): [keyCode: 53][shift: false][ctrl: true][alt: true][meta: false]
keyup (10): [keyCode: 17][shift: false][ctrl: false][alt: true][meta: false]
keyup (11): [keyCode: 17][shift: false][ctrl: false][alt: true][meta: false]
keyup (12): [keyCode: 18][shift: false][ctrl: false][alt: false][meta: false]
keyup (13): [keyCode: 18][shift: false][ctrl: false][alt: false][meta: false]
```

### Check what we detect
> Note: Please restart or reload window in VSCode after changing keyboard layout and make sure you're on the latest version

* Go to `Preferences > Keyboard Shortcuts`
* Click on the "Define Keybinding"
* Press `AltGr+5`
* We should be detecting it as `ctrl+alt+5`
![altgr-5](https://cloud.githubusercontent.com/assets/5047891/19797770/5dddc898-9ced-11e6-860f-cf9890cccfce.gif)

### Check who binds this keybinding
* In the Default Settings on the left, search for "ctrl+alt+5"
* In the User Settings on the right, search for "ctrl+alt+5"
* Please let me know if you find a keybinding bound to that key combination

Please remember to do all of the above on OSX, in the French azerty kb layout (so we're talking about the same things :))

Thanks!

**Comment by jbpotonnier** [comment]
@alexandrudima 
Here is a screenshot of the keyboard events page after pressing altGr + 5 :
![keys](https://cloud.githubusercontent.com/assets/368382/19833031/21d644c2-9e34-11e6-9926-7c0fb318b2b1.jpg)

Here is another one in vscode, after pressing altGr + 5 
![shortcuts](https://cloud.githubusercontent.com/assets/368382/19833062/a4893694-9e35-11e6-8417-a6032bc3c379.jpg)

With altGr pressed, the keys from 1 to 5 look dead, no event.

**Comment by alexandrudima** [comment]
@jbpotonnier Given your first screenshot, it looks like this is a Chromium issue. i.e. Chromium gives a `keydown` and `keyup` event only for `keyCode: 18` (`Alt`).

In other words, we are not receiving `AltGr+5` because Chromium is not giving it to us.

You might have some software installed on your system that eats up these key combinations at the OS level or this might be an issue in Chromium. Please create an issue on Chromium at https://bugs.chromium.org/p/chromium/issues/list and link to it here.

**Comment by HookyQR** [comment]
It's caused by VS Code's attempt to 're-map' the keys codes so the shortcuts map to the QWERTY key positions rather than the characters. Try the hack in https://github.com/Microsoft/vscode/issues/15624#issuecomment-261412779, and if that helps, perhaps comment on the PR https://github.com/Microsoft/vscode/pull/15687.

I can use a French keyboard (just tested it) with the 're-mapping' removed.

**Comment by Overcyn** [comment]
@HookyQR That finally fixes the issue for me!

**Comment by alexandrudima** [comment]
I have validated that we now (i.e. after PR #22894 aka #17521 -- in today's upcoming Insiders build) appear to behave correctly for this keyboard layout

![screen shot 2017-03-24 at 11 10 57](https://cloud.githubusercontent.com/assets/5047891/24289676/963c3ae0-1082-11e7-8cf0-75145fe0d798.png)

I pressed `` alt+[Digit5] `` (aka `` alt+( `` and I correctly got `{` typed in the editor.

---

This makes me believe the fix in PR #22894 aka #17521 resolves this issue.
